File: native/windows/Services/ExternalVideoRecorder.cs
using System.IO;
using System.Text;

namespace NikonLink.Windows.Services;

/// <summary>
/// Streams JPEG live-view frames into a seekable Motion-JPEG AVI. This keeps
/// tethered recording independent of vendor-specific movie download APIs.
/// </summary>
public sealed class ExternalVideoRecorder : IDisposable
{
    public sealed record RecordingResult(string Path, int Frames, long Bytes);

    private sealed record IndexEntry(long Offset, int Size);

    private readonly object _gate = new();
    private readonly List<IndexEntry> _index = [];
    private FileStream? _stream;
    private BinaryWriter? _writer;
    private string? _path;
    private int _frameRate;
    private int _width;
    private int _height;
    private int _largestFrame;
    private long _lastFrameTicks;
    private long _riffSizeOffset;
    private long _totalFramesOffset;
    private long _streamLengthOffset;
    private long _suggestedBufferOffset;
    private long _streamSuggestedBufferOffset;
    private long _moviSizeOffset;
    private long _moviTypeOffset;
    private long _moviDataOffset;

    public bool IsRecording
    {
        get { lock (_gate) return _path is not null; }
    }

    public void Start(string path, int frameRate)
    {
        lock (_gate)
        {
            if (_path is not null) throw new InvalidOperationException(
                "外录已经开始。");
            Directory.CreateDirectory(Path.GetDirectoryName(path) ?? ".");
            _path = path;
            _frameRate = Math.Clamp(frameRate, 1, 120);
            _width = 0;
            _height = 0;
            _largestFrame = 0;
            _lastFrameTicks = 0;
            _index.Clear();
        }
    }

    public void AppendJpeg(byte[] jpeg, bool throttle = true)
    {
        lock (_gate)
        {
            if (_path is null || jpeg.Length == 0) return;
            var now = System.Diagnostics.Stopwatch.GetTimestamp();
            if (throttle)
            {
                var minimumTicks = System.Diagnostics.Stopwatch.Frequency /
                    Math.Max(1, _frameRate);
                if (_lastFrameTicks != 0 && now - _lastFrameTicks < minimumTicks)
                {
                    return;
                }
            }
            if (_writer is null)
            {
                (_width, _height) = JpegDimensions(jpeg);
                _stream = new FileStream(
                    _path,
                    FileMode.Create,
                    FileAccess.ReadWrite,
                    FileShare.Read,
                    1024 * 1024,
                    FileOptions.SequentialScan);
                _writer = new BinaryWriter(_stream, Encoding.ASCII, leaveOpen: true);
                WriteHeader();
            }
            var chunkOffset = _stream!.Position;
            Ascii("00dc");
            Int32(jpeg.Length);
            _writer.Write(jpeg);
            if ((jpeg.Length & 1) != 0) _writer.Write((byte)0);
            _index.Add(new IndexEntry(chunkOffset - _moviTypeOffset, jpeg.Length));
            _largestFrame = Math.Max(_largestFrame, jpeg.Length);
            _lastFrameTicks = now;
        }
    }

    public RecordingResult? StopIfRecording()
    {
        lock (_gate)
        {
            return _path is null ? null : StopLocked();
        }
    }

    public RecordingResult Stop()
    {
        lock (_gate)
        {
            if (_path is null) throw new InvalidOperationException(
                "外录尚未开始。");
            return StopLocked();
        }
    }

    public void Abort()
    {
        lock (_gate)
        {
            var abandoned = _path;
            _path = null;
            _writer?.Dispose();
            _writer = null;
            _stream?.Dispose();
            _stream = null;
            _index.Clear();
            if (abandoned is not null) File.Delete(abandoned);
        }
    }

    public void Dispose() => Abort();

    private RecordingResult StopLocked()
    {
        var finished = _path!;
        _path = null;
        if (_writer is null || _stream is null || _index.Count == 0)
        {
            _writer?.Dispose();
            _stream?.Dispose();
            _writer = null;
            _stream = null;
            if (File.Exists(finished)) File.Delete(finished);
            throw new IOException("没有收到可写入的实时取景帧。");
        }
        var frameCount = _index.Count;
        var indexStart = _stream.Position;
        Ascii("idx1");
        Int32(frameCount * 16L);
        foreach (var entry in _index)
        {
            Ascii("00dc");
            Int32(0x10);
            Int32(entry.Offset);
            Int32(entry.Size);
        }
        var fileLength = _stream.Position;
        PatchInt32(_riffSizeOffset, fileLength - 8);
        PatchInt32(_totalFramesOffset, frameCount);
        PatchInt32(_streamLengthOffset, frameCount);
        PatchInt32(_suggestedBufferOffset, _largestFrame);
        PatchInt32(_streamSuggestedBufferOffset, _largestFrame);
        PatchInt32(_moviSizeOffset, 4 + indexStart - _moviDataOffset);
        _writer.Flush();
        _stream.Flush(flushToDisk: true);
        _writer.Dispose();
        _stream.Dispose();
        _writer = null;
        _stream = null;
        _index.Clear();
        return new RecordingResult(finished, frameCount, fileLength);
    }

    private void WriteHeader()
    {
        Ascii("RIFF");
        _riffSizeOffset = _stream!.Position;
        Int32(0);
        Ascii("AVI ");
        Ascii("LIST");
        var hdrlSizeOffset = _stream.Position;
        Int32(0);
        Ascii("hdrl");
        var hdrlDataOffset = _stream.Position;

        Ascii("avih");
        Int32(56);
        Int32(1_000_000 / _frameRate);
        Int32(0); Int32(0); Int32(0x10);
        _totalFramesOffset = _stream.Position;
        Int32(0);
        Int32(0); Int32(1);
        _suggestedBufferOffset = _stream.Position;
        Int32(0);
        Int32(_width); Int32(_height);
        Int32(0); Int32(0); Int32(0); Int32(0);

        Ascii("LIST");
        var strlSizeOffset = _stream.Position;
        Int32(0);
        Ascii("strl");
        var strlDataOffset = _stream.Position;
        Ascii("strh");
        Int32(56);
        Ascii("vids"); Ascii("MJPG");
        Int32(0); Int16(0); Int16(0); Int32(0);
        Int32(1); Int32(_frameRate); Int32(0);
        _streamLengthOffset = _stream.Position;
        Int32(0);
        _streamSuggestedBufferOffset = _stream.Position;
        Int32(0);
        Int32(uint.MaxValue); Int32(0);
        Int16(0); Int16(0); Int16(_width); Int16(_height);

        Ascii("strf");
        Int32(40); Int32(40);
        Int32(_width); Int32(_height);
        Int16(1); Int16(24); Ascii("MJPG");
        Int32((long)_width * _height * 3);
        Int32(0); Int32(0); Int32(0); Int32(0);
        var headerEnd = _stream.Position;
        PatchInt32(strlSizeOffset, headerEnd - strlDataOffset + 4);
        PatchInt32(hdrlSizeOffset, headerEnd - hdrlDataOffset + 4);
        _stream.Position = headerEnd;
        Ascii("LIST");
        _moviSizeOffset = _stream.Position;
        Int32(0);
        _moviTypeOffset = _stream.Position;
        Ascii("movi");
        _moviDataOffset = _stream.Position;
    }

    private void PatchInt32(long offset, long value)
    {
        var current = _stream!.Position;
        _stream.Position = offset;
        Int32(value);
        _stream.Position = current;
    }

    private void Ascii(string value) =>
        _writer!.Write(Encoding.ASCII.GetBytes(value));

    private void Int16(long value)
    {
        _writer!.Write((byte)(value & 0xff));
        _writer.Write((byte)((value >> 8) & 0xff));
    }

    private void Int32(long value)
    {
        _writer!.Write((byte)(value & 0xff));
        _writer.Write((byte)((value >> 8) & 0xff));
        _writer.Write((byte)((value >> 16) & 0xff));
        _writer.Write((byte)((value >> 24) & 0xff));
    }

    private static (int Width, int Height) JpegDimensions(byte[] jpeg)
    {
        if (jpeg.Length < 4 || jpeg[0] != 0xff || jpeg[1] != 0xd8)
        {
            throw new IOException("实时取景帧不是有效 JPEG。");
        }
        var offset = 2;
        while (offset + 8 < jpeg.Length)
        {
            while (offset < jpeg.Length && jpeg[offset] != 0xff) offset++;
            while (offset < jpeg.Length && jpeg[offset] == 0xff) offset++;
            if (offset >= jpeg.Length) break;
            var marker = jpeg[offset++];
            if (marker is 0xd8 or 0xd9) continue;
            if (offset + 1 >= jpeg.Length) break;
            var length = (jpeg[offset] << 8) | jpeg[offset + 1];
            if (length < 2 || offset + length > jpeg.Length) break;
            if (marker is >= 0xc0 and <= 0xc3 or >= 0xc5 and <= 0xc7 or
                >= 0xc9 and <= 0xcb or >= 0xcd and <= 0xcf)
            {
                var height = (jpeg[offset + 3] << 8) | jpeg[offset + 4];
                var width = (jpeg[offset + 5] << 8) | jpeg[offset + 6];
                if (width > 0 && height > 0) return (width, height);
            }
            offset += length;
        }
        throw new IOException("无法读取实时取景 JPEG 尺寸。");
    }
}

File: native/ios/NikonLink/Models/ExternalVideoRecorder.swift
import Foundation

/// Streams JPEG live-view frames to a seekable Motion-JPEG AVI without a
/// vendor movie-download API or an in-memory recording buffer.
final class ExternalVideoRecorder {
    struct Result {
        let url: URL
        let frames: Int
        let bytes: UInt64
    }

    private struct IndexEntry {
        let offset: UInt64
        let size: Int
    }

    private let lock = NSLock()
    private var handle: FileHandle?
    private var target: URL?
    private var frameRate = 30
    private var width = 0
    private var height = 0
    private var largestFrame = 0
    private var lastFrameTime: TimeInterval = 0
    private var index: [IndexEntry] = []
    private var riffSizeOffset: UInt64 = 0
    private var totalFramesOffset: UInt64 = 0
    private var streamLengthOffset: UInt64 = 0
    private var suggestedBufferOffset: UInt64 = 0
    private var streamSuggestedBufferOffset: UInt64 = 0
    private var moviSizeOffset: UInt64 = 0
    private var moviTypeOffset: UInt64 = 0
    private var moviDataOffset: UInt64 = 0

    var isRecording: Bool {
        lock.lock()
        defer { lock.unlock() }
        return target != nil
    }

    func start(at url: URL, frameRate: Double) throws {
        lock.lock()
        defer { lock.unlock() }
        guard target == nil else {
            throw RecorderError.invalidState("外录已经开始。")
        }
        try FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        target = url
        self.frameRate = max(1, min(120, Int(frameRate.rounded())))
        width = 0
        height = 0
        largestFrame = 0
        lastFrameTime = 0
        index.removeAll(keepingCapacity: true)
    }

    func append(jpeg: Data, bypassThrottle: Bool = false) throws {
        lock.lock()
        defer { lock.unlock() }
        guard target != nil, !jpeg.isEmpty else { return }
        let now = ProcessInfo.processInfo.systemUptime
        if !bypassThrottle,
           lastFrameTime > 0,
           now - lastFrameTime < 1.0 / Double(frameRate) {
            return
        }
        if handle == nil {
            (width, height) = try Self.jpegDimensions(jpeg)
            guard let target else { return }
            FileManager.default.createFile(atPath: target.path, contents: nil)
            handle = try FileHandle(forUpdating: target)
            try writeHeader()
        }
        guard let handle else { return }
        let chunkOffset = handle.offsetInFile
        writeASCII("00dc")
        writeInt32(UInt64(jpeg.count))
        handle.write(jpeg)
        if !jpeg.count.isMultiple(of: 2) {
            handle.write(Data([0]))
        }
        index.append(
            IndexEntry(
                offset: chunkOffset - moviTypeOffset,
                size: jpeg.count
            )
        )
        largestFrame = max(largestFrame, jpeg.count)
        lastFrameTime = now
    }

    func stopIfRecording() throws -> Result? {
        lock.lock()
        defer { lock.unlock() }
        return target == nil ? nil : try stopLocked()
    }

    func abort() {
        lock.lock()
        defer { lock.unlock() }
        let abandoned = target
        target = nil
        try? handle?.close()
        handle = nil
        index.removeAll(keepingCapacity: false)
        if let abandoned {
            try? FileManager.default.removeItem(at: abandoned)
        }
    }

    private func stopLocked() throws -> Result {
        guard let finished = target else {
            throw RecorderError.invalidState("外录尚未开始。")
        }
        target = nil
        guard let handle, !index.isEmpty else {
            try? self.handle?.close()
            self.handle = nil
            try? FileManager.default.removeItem(at: finished)
            throw RecorderError.invalidState("没有收到可写入的实时取景帧。")
        }
        let frameCount = index.count
        let indexStart = handle.offsetInFile
        writeASCII("idx1")
        writeInt32(UInt64(frameCount * 16))
        for entry in index {
            writeASCII("00dc")
            writeInt32(0x10)
            writeInt32(entry.offset)
            writeInt32(UInt64(entry.size))
        }
        let fileLength = handle.offsetInFile
        patchInt32(at: riffSizeOffset, value: fileLength - 8)
        patchInt32(at: totalFramesOffset, value: UInt64(frameCount))
        patchInt32(at: streamLengthOffset, value: UInt64(frameCount))
        patchInt32(at: suggestedBufferOffset, value: UInt64(largestFrame))
        patchInt32(
            at: streamSuggestedBufferOffset,
            value: UInt64(largestFrame)
        )
        patchInt32(
            at: moviSizeOffset,
            value: 4 + indexStart - moviDataOffset
        )
        handle.synchronizeFile()
        try handle.close()
        self.handle = nil
        index.removeAll(keepingCapacity: false)
        return Result(url: finished, frames: frameCount, bytes: fileLength)
    }

    private func writeHeader() throws {
        guard let handle else { return }
        writeASCII("RIFF")
        riffSizeOffset = handle.offsetInFile
        writeInt32(0)
        writeASCII("AVI ")
        writeASCII("LIST")
        let hdrlSizeOffset = handle.offsetInFile
        writeInt32(0)
        writeASCII("hdrl")
        let hdrlDataOffset = handle.offsetInFile

        writeASCII("avih")
        writeInt32(56)
        writeInt32(UInt64(1_000_000 / frameRate))
        writeInt32(0); writeInt32(0); writeInt32(0x10)
        totalFramesOffset = handle.offsetInFile
        writeInt32(0)
        writeInt32(0); writeInt32(1)
        suggestedBufferOffset = handle.offsetInFile
        writeInt32(0)
        writeInt32(UInt64(width)); writeInt32(UInt64(height))
        writeInt32(0); writeInt32(0); writeInt32(0); writeInt32(0)

        writeASCII("LIST")
        let strlSizeOffset = handle.offsetInFile
        writeInt32(0)
        writeASCII("strl")
        let strlDataOffset = handle.offsetInFile
        writeASCII("strh")
        writeInt32(56)
        writeASCII("vids"); writeASCII("MJPG")
        writeInt32(0); writeInt16(0); writeInt16(0); writeInt32(0)
        writeInt32(1); writeInt32(UInt64(frameRate)); writeInt32(0)
        streamLengthOffset = handle.offsetInFile
        writeInt32(0)
        streamSuggestedBufferOffset = handle.offsetInFile
        writeInt32(0)
        writeInt32(0xffff_ffff); writeInt32(0)
        writeInt16(0); writeInt16(0)
        writeInt16(UInt64(width)); writeInt16(UInt64(height))

        writeASCII("strf")
        writeInt32(40); writeInt32(40)
        writeInt32(UInt64(width)); writeInt32(UInt64(height))
        writeInt16(1); writeInt16(24); writeASCII("MJPG")
        writeInt32(UInt64(width * height * 3))
        writeInt32(0); writeInt32(0); writeInt32(0); writeInt32(0)
        let headerEnd = handle.offsetInFile
        patchInt32(
            at: strlSizeOffset,
            value: headerEnd - strlDataOffset + 4
        )
        patchInt32(
            at: hdrlSizeOffset,
            value: headerEnd - hdrlDataOffset + 4
        )
        handle.seek(toFileOffset: headerEnd)
        writeASCII("LIST")
        moviSizeOffset = handle.offsetInFile
        writeInt32(0)
        moviTypeOffset = handle.offsetInFile
        writeASCII("movi")
        moviDataOffset = handle.offsetInFile
    }

    private func patchInt32(at offset: UInt64, value: UInt64) {
        guard let handle else { return }
        let current = handle.offsetInFile
        handle.seek(toFileOffset: offset)
        writeInt32(value)
        handle.seek(toFileOffset: current)
    }

    private func writeASCII(_ value: String) {
        handle?.write(value.data(using: .ascii) ?? Data())
    }

    private func writeInt16(_ value: UInt64) {
        handle?.write(Data([
            UInt8(value & 0xff),
            UInt8((value >> 8) & 0xff)
        ]))
    }

    private func writeInt32(_ value: UInt64) {
        handle?.write(Data([
            UInt8(value & 0xff),
            UInt8((value >> 8) & 0xff),
            UInt8((value >> 16) & 0xff),
            UInt8((value >> 24) & 0xff)
        ]))
    }

    private static func jpegDimensions(_ data: Data) throws -> (Int, Int) {
        let bytes = [UInt8](data)
        guard bytes.count >= 4, bytes[0] == 0xff, bytes[1] == 0xd8 else {
            throw RecorderError.invalidFrame
        }
        var offset = 2
        while offset + 8 < bytes.count {
            while offset < bytes.count, bytes[offset] != 0xff { offset += 1 }
            while offset < bytes.count, bytes[offset] == 0xff { offset += 1 }
            guard offset < bytes.count else { break }
            let marker = bytes[offset]
            offset += 1
            if marker == 0xd8 || marker == 0xd9 { continue }
            guard offset + 1 < bytes.count else { break }
            let length = Int(bytes[offset]) << 8 | Int(bytes[offset + 1])
            guard length >= 2, offset + length <= bytes.count else { break }
            let isSOF = (0xc0...0xc3).contains(marker)
                || (0xc5...0xc7).contains(marker)
                || (0xc9...0xcb).contains(marker)
                || (0xcd...0xcf).contains(marker)
            if isSOF {
                let height = Int(bytes[offset + 3]) << 8
                    | Int(bytes[offset + 4])
                let width = Int(bytes[offset + 5]) << 8
                    | Int(bytes[offset + 6])
                if width > 0, height > 0 { return (width, height) }
            }
            offset += length
        }
        throw RecorderError.invalidFrame
    }
}

private enum RecorderError: LocalizedError {
    case invalidFrame
    case invalidState(String)

    var errorDescription: String? {
        switch self {
        case .invalidFrame: return "无法读取实时取景 JPEG 尺寸。"
        case .invalidState(let message): return message
        }
    }
}
